src/Simplex/Chat.hs

```diff
@@ -1362,24 +1367,45 @@ processAgentMessage (Just user@User {userId, profile}) agentConnId agentMessage
           incognitoProfile <- forM customUserProfileId $ \profileId -> withStore (\db -> getProfileById db userId profileId)
           let profileToSend = fromLocalProfile $ fromMaybe profile incognitoProfile
           saveConnInfo conn connInfo
-          allowAgentConnection conn confId $ XInfo profileToSend
+          -- [async agent commands] no continuation needed, but command should be asynchronous for stability
+          allowAgentConnectionAsync user conn confId $ XInfo profileToSend
         INFO connInfo ->
           saveConnInfo conn connInfo
         MSG meta _msgFlags msgBody -> do
-          _ <- saveRcvMSG conn (ConnectionId connId) meta msgBody
-          withAckMessage agentConnId meta $ pure ()
-          ackMsgDeliveryEvent conn meta
+          cmdId <- createAckCmd conn
+          _ <- saveRcvMSG conn (ConnectionId connId) meta msgBody cmdId
+          withAckMessage agentConnId cmdId meta $ pure ()
         SENT msgId ->
           -- ? updateDirectChatItemStatus
           sentMsgDeliveryEvent conn msgId
+        OK ->
+          -- [async agent commands] continuation on receiving OK
+          withCompletedCommand conn agentMsg $ \CommandData {cmdFunction, cmdId} ->
+            when (cmdFunction == CFAckMessage) $ ackMsgDeliveryEvent conn cmdId
         MERR _ err -> toView . CRChatError $ ChatErrorAgent err -- ? updateDirectChatItemStatus
         ERR err -> toView . CRChatError $ ChatErrorAgent err
         -- TODO add debugging output
         _ -> pure ()
       Just ct@Contact {localDisplayName = c, contactId} -> case agentMsg of
+        INV (ACR _ cReq) ->
+          -- [async agent commands] XGrpMemIntro continuation on receiving INV
+          withCompletedCommand conn agentMsg $ \_ ->
+            case cReq of
+              directConnReq@(CRInvitationUri _ _) -> do
+                contData <- withStore' $ \db -> do
+                  setConnConnReqInv db user connId cReq
+                  getXGrpMemIntroContDataDirect db user ct
+                forM_ contData $ \XGrpMemIntroContData {groupId, groupMemberId, memberId, connReq, hostConnId} ->
+                  forM_ connReq $ \groupConnReq -> do
+                    hostConn <- withStore $ \db -> getConnectionById db user hostConnId
+                    let msg = XGrpMemInv memberId IntroInvitation {groupConnReq, directConnReq}
+                    void $ sendDirectMessage hostConn msg (GroupId groupId)
+                    withStore' $ \db -> updateGroupMemberStatusById db userId groupMemberId GSMemIntroInvited
+              CRContactUri _ -> throwChatError $ CECommandError "unexpected ConnectionRequestUri type"
```

Comment:
this looks almost like below

Reply:
yes, I started refactoring it initially but it either requires signalling which connection connection request refers to (direct or group) if you start from Maybe XGrpMemIntroContData, or only refactoring last 4 lines (from reading `hostConn` to status update). Starting even higher is even more hassle since store method is different. Overall I'm not sure it's worth it